Add ccoctl support to create OIDC endpoint with private S3 bucket

akhil-rane opened the pull request:
We use public S3 to store OIDC cofiguration files and use the S3 URL
as public OIDC endpoint. Some users want alternative option of having 
private S3 bucket. This commit provides an option to do that using ccoctl. 
We now have OIDC configuration files stored in private S3 bucket which 
can be accessed by IAM identity provider though public CloudFront 
distribution URL.

Ccoctl command is as follows:

```
ccoctl aws create-all --name test-cluster-name --region us-east-1 --credentials-requests-dir ./credreqs --create-private-s3-bucket
```

x-ref: https://issues.redhat.com/browse/CCO-222

akhil-rane commented:
/assign @2uasimojo @abutcher

2uasimojo commented on pkg/cmd/provisioning/aws/create_identity_provider.go:
```diff
@@ -319,12 +374,133 @@ func createOIDCBucket(client aws.Client, bucketName, name, region, targetDir str
+					return "", errors.Wrapf(err, "failed to create CloudFront origin access identity")
+				}
+				originAccessIdentityID := *createOriginAccessIdentityOutput.CloudFrontOriginAccessIdentity.Id
+				log.Printf("CloudFront origin access identity created with ID %s, waiting for 30 seconds for it to become active", originAccessIdentityID)
```
nit:
```suggestion
				log.Printf("CloudFront origin access identity created with ID %s, waiting for %s for it to become active", originAccessIdentityID, cloudFrontOriginAccessIdentityActivationGracePeriod)
```
...in case the value of that const changes.

2uasimojo commented on pkg/cmd/provisioning/aws/delete.go:
```diff
@@ -76,7 +78,100 @@ func deleteOIDCBucket(client aws.Client, bucketName, namePrefix string) error {
+			if err != nil {
+				return errors.Wrapf(err, "failed to delete the CloudFront origin access identity with ID %s", *originAccessIdentity.Id)
+			}
+			log.Printf("CloudFront origin access identity with ID  %s deleted", *originAccessIdentity.Id)
```
nit: extra space
```suggestion
			log.Printf("CloudFront origin access identity with ID %s deleted", *originAccessIdentity.Id)
```

2uasimojo reviewed (commented):
I'm too fuzzy to give this a decent review right now. I'll try to have another look Monday; but maybe we wait for Andrew :innocent:

abutcher commented on pkg/cmd/provisioning/aws/delete.go:
```diff
@@ -76,7 +78,100 @@ func deleteOIDCBucket(client aws.Client, bucketName, namePrefix string) error {
+					Id: distribution.Id,
+				})
+				if err != nil {
+					return errors.Wrapf(err, "failed to get CloudFront Distribution with ID %v", distribution.Id)
```
Nit for consistency with logging the ID.

```suggestion
					return errors.Wrapf(err, "failed to get CloudFront Distribution with ID %s", *distribution.Id)
```

abutcher commented on pkg/cmd/provisioning/aws/delete.go:
```diff
@@ -76,7 +78,100 @@ func deleteOIDCBucket(client aws.Client, bucketName, namePrefix string) error {
+					DistributionConfig: getCloudFrontDistributionOutput.Distribution.DistributionConfig,
+				})
+				if err != nil {
+					return errors.Wrapf(err, "failed to disable CloudFront Distribution with ID %v", distribution.Id)
```
```suggestion
					return errors.Wrapf(err, "failed to disable CloudFront Distribution with ID %s", *distribution.Id)
```

abutcher commented on pkg/cmd/provisioning/aws/create_identity_provider.go:
```diff
@@ -319,12 +511,133 @@ func createOIDCBucket(client aws.Client, bucketName, name, region, targetDir str
+						log.Printf("CloudFront distribution with ID %s is successfully deployed", distributionID)
+						break
+					}
+					log.Printf("Waiting for CloudFront distribution with ID %s to be deployed...", distributionID)
```
```suggestion
					log.Printf("Waiting %s for CloudFront distribution with ID %s to be deployed...", cloudFrontDistributionStatusCheckDelay, distributionID)
```

abutcher commented on pkg/cmd/provisioning/aws/delete.go:
```diff
@@ -76,7 +78,100 @@ func deleteOIDCBucket(client aws.Client, bucketName, namePrefix string) error {
+						log.Printf("CloudFront distribution with ID %s is successfully disabled", *distribution.Id)
+						break
+					}
+					log.Printf("Waiting for CloudFront distribution with ID %s to be disabled...", *distribution.Id)
```
Nit: add duration to these waiting messages.

```suggestion
					log.Printf("Waiting %s for CloudFront distribution with ID %s to be disabled...", cloudFrontDistributionStatusCheckDelay, *distribution.Id)
```

abutcher reviewed (commented):
Adding some nit comments.

akhil-rane commented:
@abutcher @2uasimojo addressed your feedback. PTAL

2uasimojo reviewed (commented):
Fine by me. Andrew for the final stamp.

akhil-rane commented:
/test e2e-aws-manual-oidc

abutcher commented:
Looks good to me
/lgtm

akhil-rane commented:
retest

akhil-rane commented:
/retest

akhil-rane commented:
/assign @jianping-shu @jeana-redhat @sferich888

akhil-rane commented:
@mtulio ccoctl work is ready. It would be great if you could review the cloudfront configuration.

jeana-redhat commented on docs/ccoctl.md:
````diff
@@ -55,11 +55,11 @@ This will write out public/private key files named `serviceaccount-signer.privat
 ```
 
-where `name` is the name used to tag and account any cloud resources that are created; `region` is the aws region in which cloud resources will be created; and `public-key-file` is the path to a public key file generated using `ccoctl aws create-key-pair` as [above](#creating-rsa-keys).
+where `name` is the name used to tag and account any cloud resources that are created; `region` is the aws region in which cloud resources will be created; and `public-key-file` is the path to a public key file generated using `ccoctl aws create-key-pair` as [above](#creating-rsa-keys). `create-private-s3-bucket` is an optional parameter which can be used to create private S3 bucket with public C ...
````
I have some concerns related to the steps in the [markdown file](https://github.com/openshift/cloud-credential-operator/blob/master/docs/sts-private-bucket.md) linked from this line. I'll follow up in Slack with you and @mtulio :slightly_smiling_face:

jeana-redhat commented:
Concerns resolved via Slack - thanks!
/label docs-approved

abutcher commented:
/lgtm

akhil-rane commented:
/retest

mtulio commented:
> @mtulio ccoctl work is ready. It would be great if you could review the cloudfront configuration.

@akhil-rane AWS resources reviewed (CloudFront Distribution, OAI, Bucket, OIDC), LGTM.

It would be nice to have to support the custom domain on CloudFront distribution - or just a reference in the product docs to the detailed steps which allow users to understand the solution and create it on their own if wanted it. The complexity of adding a custom domain on CloudFront is that it requires creating other resources, like DNS Records (OIDC endpoint and ACM RR validations) and the ACM Certificate.

/lgtm

jianping-shu commented:
The testing is ongoing.
@jeana-redhat I think we need a doc ticket for updating the sts procedure w/ create-private-s3-bucket option and the additional permissions for running the ccoctl tool. Did you create one?
@akhil-rane Need you help to provide the input for new permissions for running the ccoctl tool (it needs only iam and s3 currently, but with create-private-s3-bucket option it will need CloudFront too?)

jeana-redhat commented:
@jianping-shu yes, I agree this will need docs. I've added the docs label to acknowledge that I have capacity to document this, but I do not actually know the timeline. Will talk to Akhil about it.

akhil-rane commented:
> @akhil-rane Need you help to provide the input for new permissions for running the ccoctl tool (it needs only iam and s3 currently, but with create-private-s3-bucket option it will need CloudFront too?)

Yes, it will need CloudFront permissions.

akhil-rane commented:
> It would be nice to have to support the custom domain on CloudFront distribution - or just a reference in the product docs to the detailed steps which allow users to understand the solution and create it on their own if wanted it. The complexity of adding a custom domain on CloudFront is that it requires creating other resources, like DNS Records (OIDC endpoint and ACM RR validations) and the ACM Certificate.

@mtulio I am not aware of this stuff. I will put it on my to-do and go through it. Then we can decide on how to include it in docs.

jianping-shu commented on docs/ccoctl.md:
````diff
@@ -92,7 +92,7 @@ $ oc adm release extract --credentials-requests --cloud=aws --to=./credrequests
 
 ```bash
-$ ccoctl aws create-all --name=<name> --region=<aws-region> --credentials-requests-dir=<path-to-directory-with-list-of-credentials-requests>
+$ ccoctl aws create-all --name=<name> --region=<aws-region> --credentials-requests-dir=<path-to-directory-with-list-of-credentials-requests> 
````
Need to add optional parameter --create-private-s3-bucket here?

akhil-rane replied:
fixed

jianping-shu commented:
@akhil-rane I tested and updated with [OCP-52105](https://polarion.engineering.redhat.com/polarion/redirect/project/OSE/workitem?id=OCP-52105), PTAL!
One minor finding in the testing is "ID 30s, waiting for E32LZU2DJUUZVJ" in the log.
[cloud-user@preserve-for-hive-test run2]$ ccoctl aws create-identity-provider --name jshu-awscls --region=us-east-2 --public-key-file=serviceaccount-signer.public --create-private-s3-bucket
2022/08/26 02:42:07 Bucket jshu-awscls-oidc created
2022/08/26 02:42:08 CloudFront origin access identity created with ID 30s, waiting for E32LZU2DJUUZVJ for it to become active

akhil-rane commented:
> @akhil-rane I tested and updated with [OCP-52105](https://polarion.engineering.redhat.com/polarion/redirect/project/OSE/workitem?id=OCP-52105), PTAL!
One minor finding in the testing is "ID 30s, waiting for E32LZU2DJUUZVJ" in the log.
[cloud-user@preserve-for-hive-test run2]$ ccoctl aws create-identity-provider --name jshu-awscls --region=us-east-2 --public-key-file=serviceaccount-signer.public --create-private-s3-bucket
2022/08/26 02:42:07 Bucket jshu-awscls-oidc created
2022/08/26 02:42:08 CloudFront origin access identity created with ID 30s, waiting for E32LZU2DJUUZVJ for it to become active

@jianping-shu I have fixed the logging issue that you pointed out. The test case looks good to me

jianping-shu commented:
/label qe-approved

abutcher commented:
/lgtm

davemulford commented:
/label px-approved

abutcher commented:
/cherry-pick release-4.11
